Solution

CarlosUvaSilva opened the pull request.

david-lovelystay commented on .env:
Hello Carlos!

I'm David, a developer from the LovelyStay team. Thank you very much for submitting your test.
I'll be leaving some comments, but feel free to discuss them openly as these are my suggestions :)

--

In this case it's ok, but be wary of commiting a `.env`. especially since you're already committing a `.env.example`

CarlosUvaSilva replied:
I left this here as "proof" of the `psql` commands output.
Should have clarified this with a comment

david-lovelystay commented on .gitignore:
```diff
@@ -0,0 +1 @@
+node_modules
```
It's good to also include the output folder, even if the project is compiling the `.ts` files in runtime.
Regardless it's best practice to compile and then run, adding the compiled `.js` files destination path to the `.gitignore`.

CarlosUvaSilva replied:
I understand. I don't have much experience with TS and the little I have is using React which compiles everything anyway.
I'll add this in the next PR

david-lovelystay commented on db/fetch_users.ts:
```diff
@@ -0,0 +1,61 @@
+  languages: string | null
+) {
+  try {
+    // Giving errors with empty locations and the list filter
```
Please avoid committing commented dead code

david-lovelystay commented on db/fetch_users.ts:
```diff
@@ -0,0 +1,61 @@
+export async function fetchUsers(
+  location: string | null,
+  languages: string | null
+) {
```
I am aware that you did leave some `any` types due to time constraints, but the users that are returned here are a database table, so it shouldn't be too hard to add an interface for it with the correct types.

david-lovelystay commented on db/fetch_users.ts:
```diff
@@ -0,0 +1,61 @@
+        JOIN programming_languages pl ON ul.language_id = pl.id
+      `;
+      const languages_array = languages.split(',').map(lang => lang.trim());
+      conditions.push(`pl.name = ANY($${conditions.length + 1}::text[])`);
```
This is an over-complication of `pg-promise` parametrization.
Please re-read the documentation and try to find a better way of doing this.

I can give 2 tips in this regard:
Firstly, `pg-promise` supports other types of parameter indexing that is not positional.
Secondly, there is native array support, though it is not mandatory and can be rather confusing.

david-lovelystay commented on db/insert_repos.ts:
```diff
@@ -0,0 +1,34 @@
+          INSERT INTO programming_languages (name)
+          VALUES ($1)
+          ON CONFLICT (name)
+          DO UPDATE SET name = EXCLUDED.name
```
Isn't this basically the same as doing nothing?
If there's a conflict on name, then update name is redundant, right?

CarlosUvaSilva replied:
yes and no. if you don't update then nothing is returned even if you do `RETURNING id`
this way the name is "updated" and the id is returned. otherwise you'd need to do a separate `SELECT` statement fetching the record if the result of the original return was `null`

david-lovelystay commented on main.ts:
```diff
@@ -0,0 +1,67 @@
+  .command('get-user <username>')
+  .description('Fetches a GitHub user profile')
+  .action(async (username: string) => {
+    try {
```
Perhaps these actions functions can be declared outside of the `main` file.
It seems somewhat verbose for what is usually placed in these files.

CarlosUvaSilva commented:
hey @david-lovelystay 

I've pushed new commits addressing the feedback given.
I was able to figure out named parameters for `pg-promise` and also added some typing across the project.
Also refactored the main file so it's only responsibility is parsing command-line commands and flags.

Hope these changes make the code better and worthy of being approved.
